cpp/include/hf/flexpolyline.h

```diff
@@ -4,302 +4,465 @@
 
-namespace hf {
-
-const uint8_t FORMAT_VERSION=1;
+private:
+  explicit Precision(uint8_t value) : m_value(value) {}
+  uint8_t m_value;
+};
 
-enum class ThirdDim: int {
-    ABSENT = 0,
-    LEVEL = 1,
-    ALTITUDE = 2,
-    ELEVATION = 3,
-    /* Reserved values should not be selectable */
-    CUSTOM1 = 6,
-    CUSTOM2 = 7
+/// Informs about the type of the 3rd dimension of a 3D coordinate vector
+enum class Type3d {
+  /// E.g. floor of a building
+  LEVEL = 1,
+  /// E.g. altitude (in the air) relative to ground level or mean sea level
+  ALTITUDE = 2,
+  /// E.g. elevation above mean-sea-level
+  ELEVATION = 3,
+  /// Reserved for future types
+  RESERVED1 = 4,
+  /// Reserved for future types
+  RESERVED2 = 5,
+  /// Reserved for custom types
+  CUSTOM1 = 6,
+  /// Reserved for custom types
+  CUSTOM2 = 7,
 };
 
-template<typename Iter>
-std::string polyline_encode(Iter iter, int precision=5, ThirdDim third_dim=ThirdDim::ABSENT, int third_dim_precision=0);
+/// 2-dimensional polyline
+struct Polyline2d {
+  /// List of 2D coordinates making up this polyline
+  std::vector<std::tuple<double, double>> coordinates;
+  /// Precision of the coordinates (e.g. used for encoding,
+  /// or to report the precision supplied in encoded data)
+  Precision precision2d = *Precision::from_u32(7);
+
+  Polyline2d() = default;
+  Polyline2d(std::vector<std::tuple<double, double>> coordinates, Precision precision2d)
+      : coordinates(std::move(coordinates)), precision2d(precision2d) {}
+};
 
-template<typename Fn>
-bool polyline_decode(const std::string &encoded, Fn&& push);
+/// 3-dimensional polyline
+struct Polyline3d {
+  /// List of 3D coordinates making up this polyline
+  std::vector<std::tuple<double, double, double>> coordinates;
+  /// Precision of the 2D part of the coordinates (e.g. used for encoding,
+  /// or to report the precision supplied in encoded data)
+  Precision precision2d = *Precision::from_u32(7);
+  /// Precision of the 3D part of the coordinates (e.g. used for encoding,
+  /// or to report the precision supplied in encoded data)
+  Precision precision3d = *Precision::from_u32(3);
+  /// Type of the 3D component
+  Type3d type3d = Type3d::ELEVATION;
+
+  Polyline3d() = default;
+  Polyline3d(std::vector<std::tuple<double, double, double>> coordinates, Precision precision2d,
+             Precision precision3d, Type3d type3d)
+      : coordinates(std::move(coordinates)), precision2d(precision2d), precision3d(precision3d),
+        type3d(type3d) {}
+};
 
-ThirdDim get_third_dimension(const std::string &encoded);
+/// 2- or 3-dimensional polyline
+using Polyline = std::variant<Polyline2d, Polyline3d>;
+
+inline std::string to_string(const Polyline &polyline, std::optional<int> precision = {}) {
+  std::ostringstream out;
+  out << std::fixed;
+  std::visit(
+      [&](auto &&arg) {
+        using T = std::decay_t<decltype(arg)>;
+        if constexpr (std::is_same_v<T, Polyline2d>) {
```

Comment:
This is pretty much a 1:1 replacement for Rust's match function. It also seems to be in C++ now (at least it is almost 1:1 what is recommended in https://en.cppreference.com/w/cpp/utility/variant/visit). As always ergonomics in C++ seem to be a bit ... worse ... than in other languages :-)

`std::visit` like this should also be safe: You get a compile-time error in case you do not cover one of the variants.